### PR TITLE
Reorder ResNetV2 preactivation after branching

### LINKED ISSUE #15167: ResNetV2 implementation problem
**System information**
(not related to installation)
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Windows 10
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.5.0 GPU
- Python version: 3.8

**Describe the expected behavior**
As shown in the image below from **Identity Mappings in Deep Residual Networks** by He et al.\
ResNetV2 block has its preactivation **after** branching (bn-relu-conv-bn-relu-conv are all in contained in residual)
image
the **only** exception on preactivation is the first block of the first stage\
quote from the original paper:
> For the first Residual Unit (that follows a stand-alone convolutional layer,
> conv1), we adopt the first activation right after conv1 and before splitting into
> two paths

**Describe the current behavior**
ResNetV2 Falsefully put preactivation before branching when channel increases for **all stages**
image
As shown in above image, the output of `ResNet50V2(weights=None).summary()`\
the red arrow I added points to the shortcut path, it is connected to preact_relu, which means the preactivation is done **before** branching to residual and shortcut, which is not true according to the original paper\
a simple drawing representing the first block of each stage in current implementation, which is obviously not equivalent to the image from the original paper:
image

### PR DESCRIPTION
**Pull request by MaanasArora**
To fix #15167; reordered the layers. This won't affect the first layer declared in the ResNet class (as expected). Thanks for reviewing!

### CODE REVIEW

**Comment by qlzh727** [comment]
Note that this PR change the topology of the network, and will affect the weights/metrics of the model. Could you provide the any verification for this PR? (eg the model is achieving the expected performance with this PR), and also does it need to generate a new set of weights?

**Comment by MaanasArora** [comment]
I don't have the resources to verify the official benchmarks, but I trained ResNetV250 before and after the PR on the imagenette dataset for one epoch and the modified model performed better than the previous one at least in this epoch:

image
image
image
image

Here's a [gist](https://colab.research.google.com/gist/MaanasArora/8840ae99216c70ba21c6fd50c20d6e4e/resnet-validate.ipynb) in case you'd like to replicate or extend the results. You need to use Keras to export and upload the (uncompiled) model file after the PR.

Yes, I think we'd have to generate a new set of weights for this change, since the shortcut is not using the outputs of the preact layer anymore. I think it's worth it for consistency.

**Comment by fchollet** [comment]
We need to run inference with the pretrained weights on the ImageNet validation set and check that the metrics look good.

**Comment by MaanasArora** [comment]
@fchollet I tried running with the pretrained imagenet weights and it didn't perform well. I think since the topology of the model has changed, would we need to generate a new set of weights? Thanks

**Comment by fchollet** [comment]
The ResNetV2 weights were obtained from the official implementation IIRC, so if the weights work with the current implementation that strongly indicates that the current implementation matches the official implementation. Can you check out the original ResNetV2 implementation from the authors?

**Comment by MaanasArora** [comment]
@fchollet I checked the original repository in Lua and it seems that they have a structure similar to what we have after the PR:

```Lua
      if nInputPlane == nOutputPlane then -- most Residual Units have this shape      
         local convs = nn.Sequential()
         -- conv1x1
         convs:add(SBatchNorm(nInputPlane))
         convs:add(ReLU(true))
         convs:add(Convolution(nInputPlane,nBottleneckPlane,1,1,stride,stride,0,0))
        
         -- conv3x3
         convs:add(SBatchNorm(nBottleneckPlane))
         convs:add(ReLU(true))
         convs:add(Convolution(nBottleneckPlane,nBottleneckPlane,3,3,1,1,1,1))
        
         -- conv1x1
         convs:add(SBatchNorm(nBottleneckPlane))
         convs:add(ReLU(true))
         convs:add(Convolution(nBottleneckPlane,nOutputPlane,1,1,1,1,0,0))
        
         local shortcut = nn.Identity()
        
         return nn.Sequential()
            :add(nn.ConcatTable()
               :add(convs)
               :add(shortcut))
            :add(nn.CAddTable(true))
```
([https://github.com/KaimingHe/resnet-1k-layers/blob/master/resnet-pre-act.lua](url))

The last few lines have the entire stack and the shortcut in parallel, which means the preactivation is done after branching out.

Also, I plotted the Keras implementation before and after the PR:

**Comment by MaanasArora** [comment]
@fchollet Yes, it seems that the [official implementation in Lua](https://github.com/KaimingHe/resnet-1k-layers/blob/master/resnet-pre-act.lua) does preactivate _before_ branching every time dimensions are increased, so the current implementation would match the official implementation. Although this would be inconsistent with the original paper according to the issue report, I think we shouldn't deviate from the official implementation. I'll move this discussion to the original issue if you'd like. Thanks.

**Comment by fchollet** [comment]
> But this would be inconsistent with the paper according to the issue report, and I wonder if we should follow the paper more closely

This is a common issue. Papers describe what was intended by the authors, but the code describes what they actually tried. Often, certain APIs introduce common gotchas that make it very hard to achieve what you actually intended in your implementation. This is why your choice of API as a researcher matters a lot.

I think we can close this issue; the source of truth for the model should be its official code.
